squire_core/src/state.rs

```diff
@@ -71,17 +105,38 @@ impl ServerState for AppState {
     }
 
     async fn get_tourn(&self, id: TournamentId) -> Option<TournamentManager> {
-        let mut cursor = self.get_tourns().find(None, None).await.unwrap();
-        while let Some(tourn) = cursor.try_next().await.unwrap() {
-            if tourn.id == id {
-                return Some(tourn);
-            }
-        }
-        None
+        self.get_tourns()
+            .find_one(Some(Self::make_query(id)), None)
+            .await
+            .ok()
+            .flatten()
     }
 
     async fn persist_tourn(&self, tourn: &TournamentManager) -> bool {
-        todo!()
+        // There appears to be a problem in bson right now where `Collection::replace_one` uses the
+        // normal document serializer, but `Collection::find_one` (and `Collection::insert_one` as
+        // well) use the raw document serializer, which unfortunately behave differently. Therefore
+        // `Collection::update_one` is used as a workaround so that we can call the raw document
+        // serializer here
+        let doc: Document = bson::to_raw_document_buf(tourn)
+            .unwrap()
+            .try_into()
+            .unwrap();
+        match self
+            .get_tourns()
+            .update_one(
+                Self::make_query(tourn.id),
+                UpdateModifications::Document(doc! {"$set": doc}),
+                UpdateOptions::builder().upsert(true).build(),
+            )
+            .await
+        {
+            Ok(result) => result.matched_count != 0,
+            Err(e) => {
```

Comment:
Do we know what would trigger this? Does the returned error tell us anything? Obviously, we can't persist if we have lost our connection to the DB, but if the error is something like "document not found", we probably want to just insert the document.

Reply:
I believe this will insert the document if it is not found, since the `upsert` field is specified for the options.

Reply:
Can we please verify this with a test? I looked at the error type that the update method returns and it's not very enlightening.

Reply:
I think the `insert_fetch_tourn` test can verify, since it clears the database before calling persist. The test succeeds, and I'm able to find the document after the test runs by inspecting the database directly.

Reply:
Hmm, ya. That's a good point.